pkg/controller/master/upgrade/image_controller.go

```diff
@@ -0,0 +1,47 @@
+package upgrade
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	ctlharvesterv1 "github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1"
+)
+
+type vmImageHandler struct {
+	namespace     string
+	upgradeClient ctlharvesterv1.UpgradeClient
+	upgradeCache  ctlharvesterv1.UpgradeCache
+}
+
+func (h *vmImageHandler) OnChanged(key string, image *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
+	if image == nil || image.DeletionTimestamp != nil || image.Labels == nil || image.Namespace != upgradeNamespace || image.Labels[harvesterUpgradeLabel] == "" {
+		return image, nil
+	}
+
+	upgradeControllerLock.Lock()
+	defer upgradeControllerLock.Unlock()
+
+	upgrade, err := h.upgradeCache.Get(upgradeNamespace, image.Labels[harvesterUpgradeLabel])
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return image, nil
+		}
+		return nil, err
+	}
+
+	if harvesterv1.ImageImported.GetStatus(image) == string(corev1.ConditionTrue) {
```

Comment:
Can we use `harvesterv1.ImageImported.IsTrue(image)`?

Reply:
Fixed in https://github.com/harvester/harvester/pull/1635/commits/02f43bdf63e555b32abe7984a5680cef19f44bcc#diff-15276b9422a58603cacc38542d540cf52d04053a27b5e9dadd702b49d1cf9435R37-R52
I also include the status report if image importing fails.